aea/agent_loop.py

```diff
@@ -0,0 +1,400 @@
+            self._loop,
+        )
+        self._behaviours_registry[behaviour] = periodic_caller
+        periodic_caller.start()
+
+    def _register_all_behaviours(self) -> None:
+        """Register all AEA behaviours to run periodically."""
+        for behaviour in self._agent._get_active_behaviours():
+            self._register_behaviour(behaviour)
+
+    def _unregister_behaviour(self, behaviour: Behaviour) -> None:
+        """
+        Unregister periodic execution of the behaviour.
+
+        :param behaviour: Behaviour to schedule periodic execution.
+        :return: None
+        """
+        periodic_caller = self._behaviours_registry.pop(behaviour, None)
+        if not periodic_caller:
+            return
+        periodic_caller.stop()
+
+    def _stop_all_behaviours(self) -> None:
+        """Unregister periodic execution of all registered behaviours."""
+        for periodic_caller in self._behaviours_registry.values():
+            periodic_caller.stop()
+        self._behaviours_registry = {}
+
+    def _create_tasks(self):
+        """Create tasks to execute and wait."""
+        tasks = self._create_processing_tasks()
+        stopping_task = create_task(self._state.wait(["stopping", "error"]))
+
+        tasks.append(stopping_task)
+        return tasks
+
+    async def _cancel_and_wait_tasks(self, tasks: List[Task]) -> None:
+        """Cancel all tasks and wait they completed."""
+        for t in tasks:
+            t.cancel()
+
+        await asyncio.wait(tasks, return_when=ALL_COMPLETED)
+
+        for t in tasks:
+            if t.cancelled():
+                continue
+            exc = t.exception()
+            if exc:
+                self._exceptions.append(cast(Exception, exc))
+
+    async def _run(self) -> None:
+        """Run all tasks and wait for stopping state."""
+        tasks = self._create_tasks()
+
+        await asyncio.wait(tasks, return_when=FIRST_COMPLETED)
+
+        # one task completed by some reason: error or state == stopped.
+        # clean up
+        self._stop_all_behaviours()
+        await self._cancel_and_wait_tasks(tasks)
+
+        if self._exceptions:
+            # check exception raised during run
+            raise AgentLoopException(self._exceptions)
+
+        self._state.set("stopped")
+
+    def _create_processing_tasks(self) -> List[Task]:
+        """
+        Create processing tasks.
+
+        :return: list of asyncio Tasks
+        """
+        tasks = [
+            self._task_process_inbox(),
+            self._task_process_internal_messages(),
+            self._task_process_new_behaviours(),
+        ]
+        return list(map(create_task, tasks))
+
+    @property
+    def is_running(self) -> bool:
+        """Get running state of the loop."""
+        return self._state.state == "started"
+
+    async def _task_process_inbox(self) -> None:
+        """Process incoming messages."""
+        queue = self._agent._inbox
```

Comment:
minor, optional, for readability: rename it as `inbox`, because it's an `InBox` instance. Maybe also add some type annotation.

Reply:
make sense, thanks